### PR TITLE
Ticket/aotech 6795 arbitrary cli options

### PR DESCRIPTION
**Pull request by phatsk**
This Pull Request allows options to be passed to CLI that are currently only configurable in the UI.

For example, now you can run

```
wp press-sync posts --ps_preserve_ids=true --ps_delta_date="2018-02-01" --ps_partial_terms=true
```

There is also a preliminary `--verbose` option that will attempt to log the result of API requests to the console.

### CODE REVIEW

**Comment by prometheas** [comment]
Underscores don't really befit option names in CLI "UIs". I recommend switching to hyphens, such that `--ps_delta_date="2018-02-01"` becomes `--ps-delta-date="2018-02-01"`

**Review by prometheas — APPROVED** [review]
💥

File: includes/class-api.php [hunk 1]
```diff
@@ -176,13 +176,13 @@ public function validate_sync_key() {
...
+		$objects_to_sync         = $request->get_param( 'ps_objects_to_sync' );
 		$objects                 = $request->get_param( 'objects' );
-		$duplicate_action        = $request->get_param( 'duplicate_action' ) ? $request->get_param( 'duplicate_action' ) : 'skip';
-		$force_update            = $request->get_param( 'force_update' );
-		$this->skip_assets       = (bool) $request->get_param( 'skip_assets' );
-		$this->preserve_ids      = (bool) $request->get_param( 'preserve_ids' );
-		$this->fix_terms         = (bool) $request->get_param( 'fix_terms' );
+		$duplicate_action        = $request->get_param( 'ps_duplicate_action' ) ? $request->get_param( 'ps_duplicate_action' ) : 'skip';
+		$force_update            = $request->get_param( 'ps_force_update' );
+		$this->skip_assets       = 'true' == $request->get_param( 'ps_skip_assets' );
```
Comment (jmichaelward) [inline includes/class-api.php]: Why not `===` strict comparison?

File: includes/class-cli.php [hunk 2]
```diff
@@ -158,4 +204,37 @@ private function return_response( $response = array() ) {
...
+	 *
+	 * @since NEXT
+	 * @param  array $assoc_args The associative args from the command line.
+	 * @return array
+	 */
+	private function parse_assoc_args( $assoc_args ) {
+		$valid_keys    = array_keys( $this->plugin->parse_sync_settings() );
+		$valid_keys[]  = 'verbose';
+		$prepared_args = [];
+
```
Comment (jmichaelward) [inline includes/class-cli.php:234]: So, I get from the title that we're "parsing the associative arguments", but what does that actually mean to a developer who might look at this code later? Can we provide an example input and an example output in the doc block so it's more clear what kinds of things are being stripped out of the args? Without named functions processing these foreach loops, we're just kind of left knowing that certain kinds of args are getting excluded.

**Review by rakeshdx — APPROVED** [review]

File: includes/class-press-sync.php [hunk 3]
```diff
@@ -1072,25 +1064,44 @@ public function order_to_sync_all( $order_to_sync_all = array() ) {
...
-			'remote_domain'        => get_option( 'ps_remote_domain' ),
-			'ps_remote_key'        => get_option( 'ps_remote_key' ),
-			'sync_method'          => get_option( 'ps_sync_method' ),
-			'objects_to_sync'      => get_option( 'ps_objects_to_sync' ),
-			'duplicate_action'     => get_option( 'ps_duplicate_action' ),
-			'force_update'         => get_option( 'ps_force_update', false ),
-			'skip_assets'          => get_option( 'ps_skip_assets', false ),
-			'options'              => get_option( 'ps_options_to_sync' ),
+		static $default_settings = array(
+			'ps_remote_domain'     => null,
```
Comment (jmichaelward) [inline includes/class-press-sync.php:1068]: Glad to see a unified prefixing here.

**Review by jmichaelward — APPROVED** [review]

File: includes/class-cli.php (hunk 2 above)
Comment (phatsk) [reply includes/class-cli.php:234]: I added some elaboration to the docblock for clarification.

**Comment by phatsk** [comment]
I restored the previous strict batching with a `(bool)` cast for the incoming request parameters and added a check for string "false" when options are being processed. There was an edge-case where string "false" was being passed to the CLI, and casting that to `(bool)` results in `true`.

**Pull request merged**